[16.0][FIX] auditlog: Fix Invalid Reading of Stale Database Values

BT-pgaiser opened the pull request:
Due to the introduction of the `ThrowAwayCache`, in full log mode when attempting to log a write, the reading of the '_new_' values effectively reads the '_old_' values because prior changes were not flushed to the database yet at that point. This resulted in missing log lines in the created log entry since the old vs. new value comparison did not indicate any changes.

It has been confirmed that without the introduced patch for the database flush, the added test fails as expected.

BT-pgaiser commented:
Tackles issues [#3444](https://github.com/OCA/server-tools/issues/3444) and [#3420](https://github.com/OCA/server-tools/issues/3420).

BT-pgaiser commented:
@StefanRijnhart Requesting a review.

StefanRijnhart commented on auditlog/models/rule.py:
```diff
@@ -485,6 +485,10 @@ def write_full(self, vals, **kwargs):
             if self.env.user in users_to_exclude:
                 return result
 
+            self.env[records_write._name].flush_model(
```
Maybe

```suggestion
           self.flush_recordset(
```

is even more precise?

BT-pgaiser replied:
Good point. Changed.

StefanRijnhart reviewed (commented):
Very nice, thank you! Just a nitpick regarding the flushing.

StefanRijnhart reviewed (approved):
Thanks!

StefanRijnhart commented:
Oh, please squash commits into one.

BT-pgaiser committed:
[FIX] auditlog: Fix invalid reading of stale database values.

Due to the introduction of the ThrowAwayCache, in full log mode when attempting
to log a write, the reading of the 'new' values effectively reads the 'old'
values because prior changes were not flushed to the database yet at
that point. This resulted in missing log lines in the created log entry since
the old vs. new value comparison did not indicate any changes.

BT-pgaiser commented:
@StefanRijnhart Can we please merge this if it's OK from your side? Thanks.

StefanRijnhart commented:
We cannot merge this yet, see section 6 of https://odoo-community.org/resources/code.

StefanRijnhart commented:
Dear BT members, please consider to not pile up these inter-team reviews as you did here. It feels like you are playing the system. It would be different if the reviews come from known and appreciated contributors to the OCA but I see that for all of you the review on this PR is the only OCA review you did in the whole month (or, ever).

BT-rmartin commented:
@StefanRijnhart 

According to the rules...
_Your code will be merged upon 3 positive reviews within 5 days (or 2 after more than 5 days). At least one of the review above must be from a member of the [PSC](https://odoo-community.org/page/List) or someone from the [OCA Core Maintainer](https://github.com/orgs/OCA/teams/core-maintainers). Find more info [here](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#review)._

So max 2 reviews can be done by whoever is not PSC or OCA Core Maintainer, so I guess here is the problem that we did approve it 4 times. 
 
I want to clarify our position. We don't just simply "approve", we do take the time to properly review the code, and we all are OCA members, although perhaps not with huge activity.  
In particular @BT-anieto has contributed with several fixes or migrations of modules 
https://github.com/pulls?q=is%3Apr+author%3ABT-anieto+archived%3Afalse+user%3AOCA+is%3Aclosed
Also @BT-tkarpinski did a migration and a fix contributed to OCA
https://github.com/pulls?q=is%3Apr+author%3ABT-tkarpinski+archived%3Afalse+is%3Aclosed+user%3AOCA
In the case of @BT-ojossen he has also some contributions with new modules, fixes or translations https://github.com/pulls?q=is%3Apr+author%3ABT-ojossen+archived%3Afalse+is%3Aclosed+user%3AOCA+
Finally, @BT-ssteiner also did a migration and an improvement in the past https://github.com/pulls?q=is%3Apr+author%3ABT-ssteiner+archived%3Afalse+is%3Aclosed+user%3AOCA+

Our organization is not "OCA-first", but we do use OCA modules and contribute mainly with migrations, improvements and fixes from time to time, as well as reviews of the contributions of our colleagues. 

My question is: Is it a problem that we contribute even if our contributions are not monthly or very frequent?
I would expect that any contributio, whether code, reviews, or support, is welcome.

So, what is the expectation from the OCA regarding reviewers?
Should we limit ourselves to one reviewer per company?
Is it a requirement to review PRs from other contributors as well?

We want to do it better and be more aligned with community expectations. 
Thanks in advance for your insights.

dreispt commented:
/ocabot merge patch

StefanRijnhart commented:
For the record, no port is required, at least not to 18.0, as the issue surfaced there in an earlier stage and the flush was already applied (https://github.com/OCA/server-tools/blob/2c75742/auditlog/models/rule.py#L462). I tested with this PRs additional test method and the test succeeds without this PRs additional code change. If the flush would be removed in the future for some reason, it would lead to the failure of test TestAuditlogFull.test_update_logs_have_lines_with_changes.
